### PR TITLE
Add vector types.

### PR DESCRIPTION
**Pull request by mzient**
Vector types with shader-like usage.
Example:
```
ivec3 a = { 1, 2, 3 };
vec3 b = a + 1.0f;  // == vec3{2, 3, 4} 
vec4 c = cat(b, 5.0f);  // == vec4{2, 3, 4, 5};
vec3 d = a - b;  // == vec3{-1, -1, -1};
```

Partial review for #981.

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: include/dali/core/geom/vec.h [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+#else
+  return std::sqrt(length_square());
+#endif
+  }
+  DALI_HOST_DEV inline vec normalized() const {
+    auto lsq = length_square();
+    return *this * rsqrt(lsq);
+  }
+
+  /// @brief Returns a copy. Doesn't promoto type to int.
```
Comment (klecki) [inline include/dali/core/geom/vec.h]:
```suggestion
  /// @brief Returns a copy. Doesn't promote type to int.
```

File: include/dali/core/geom/vec.h [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+
+  #define DEFINE_ASSIGN_VEC_OP(op)\
+  template <typename U>\
+  DALI_HOST_DEV vec &operator op(const vec<N, U> &rhs) {\
+    for (size_t i = 0; i < N; i++)\
+      v[i] op rhs[i];\
+    return *this;\
+  }\
+  template <typename U>\
+  DALI_HOST_DEV std::enable_if_t<!is_vec<U>::value, vec &> operator op(const U &rhs) {\
```
Comment (klecki) [inline include/dali/core/geom/vec.h]: Shouldn't this be only allowed for scalars?

File: include/dali/core/geom/vec.h [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+#define DEFINE_ELEMENTIWSE_VEC_BIN_OP(op)\
+template <size_t N, typename T, typename U>\
+DALI_HOST_DEV inline auto operator op(const vec<N, T> &a, const vec<N, U> &b) {\
+  vec<N, decltype(T() op U())> ret;\
+  for (size_t i = 0; i < N; i++)\
+    ret[i] = a[i] op b[i];\
+  return ret;\
+}\
+template <size_t N, typename T, typename U, typename R = decltype(T() op U())>\
+DALI_HOST_DEV inline std::enable_if_t<!is_vec<U>::value, vec<N, R>> \
```
Comment (klecki) [inline include/dali/core/geom/vec.h]: Scalar here as well?

**Comment by mzient** [comment]
!build

File: include/dali/core/geom/vec.h [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+/// @brief Calculates `z` coordinate of a cross product of two 2D vectors
+template <typename T, typename U>
+DALI_HOST_DEV
+constexpr auto cross(const vec<2, T> &a, const vec<2, U> &b) {
+  return a.x * b.y - b.x * a.y;
+}
+
+#define DEFINE_ELEMENTIWSE_VEC_BIN_OP(op)\
+template <size_t N, typename T, typename U>\
+DALI_HOST_DEV inline auto operator op(const vec<N, T> &a, const vec<N, U> &b) {\
```
Comment (klecki) [inline include/dali/core/geom/vec.h]: This has unfortunate effect of `decltype(vec<4, int16_t>() + vec<4, int16_t>()) == decltype(vec<4, int32_t>)`. 
Is that desired bahaviour?

File: include/dali/core/geom/vec.h [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+  vec<sub_n, T> ret = {};
+  for (size_t i = 0; i < sub_n; i++)
+    ret[i] = orig[i + start];
+  return ret;
+}
+
+template <typename To, size_t size, typename From,
+          typename OutputType = vec<size*sizeof(From)/sizeof(To), To>>
+DALI_HOST_DEV
+constexpr OutputType bitcast(const vec<size, From> &v) {
```
Comment (klecki) [inline include/dali/core/geom/vec.h]: I should add a CI check that searches for bitcast and links to this: https://chromium.googlesource.com/chromium/src/base/+/master/bit_cast.h
:P

**Comment by mzient** [comment]
!build

File: include/dali/core/geom/vec.h (hunk 4 above)
Comment (mzient) [reply include/dali/core/geom/vec.h]: Fixed. Now the promotion rules are as follows:
For vectors:
non-fp (op) fp -> fp
fp (op) non-fp -> fp
non-fp1 (op) non-fp2 -> error
fp1 (op) fp2 -> error
For vectors-scalar:
non-fp (op) fp -> fp
fp (op) non-fp -> fp
int_v (op) int_s -> int_s
fp_v (op) fp_s -> fp_v

**Comment by mzient** [comment]
!build

File: include/dali/core/math_util.h [hunk 6]
```diff
@@ -53,6 +56,71 @@ DALI_HOST_DEV constexpr T clamp(const T &value, const T &lo, const T &hi) {
...
+DALI_HOST_DEV inline float rsqrt(float x) {
+#ifdef __CUDA_ARCH__
+  return __frsqrt_rn(x);
+#elif defined __SSE__
+  __m128 X = _mm_set_ss(x);
+  __m128 tmp = _mm_rsqrt_ss(X);
+  float y = _mm_cvtss_f32(tmp);
+  return y * (1.5f - x*0.5f * y*y);
+#else
+  int32_t i;
```
Comment (JanuszL) [inline include/dali/core/math_util.h]: Any reference to source of that magic?

File: include/dali/core/math_util.h [hunk 7]
```diff
@@ -53,6 +56,71 @@ DALI_HOST_DEV constexpr T clamp(const T &value, const T &lo, const T &hi) {
...
+
+DALI_HOST_DEV inline float fast_rsqrt(float x) {
+#ifdef __CUDA_ARCH__
+  return __frsqrt_rn(x);
+#elif defined __SSE__
+  __m128 X = _mm_set_ss(x);
+  __m128 tmp = _mm_rsqrt_ss(X);
+  return _mm_cvtss_f32(tmp);
+#else
+  int32_t i;
```
Comment (JanuszL) [inline include/dali/core/math_util.h]: Any reference to source of that magic?

File: include/dali/core/math_util.h [hunk 8]
```diff
@@ -53,6 +56,71 @@ DALI_HOST_DEV constexpr T clamp(const T &value, const T &lo, const T &hi) {
...
+DALI_HOST_DEV inline double rsqrt(double x) {
+  return 1.0/sqrt(x);
+}
+
+DALI_HOST_DEV inline double fast_rsqrt(double x) {
+#ifdef  __CUDA_ARCH__
+  // not likely to be used at device side anyway
+  return fast_rsqrt(static_cast<float>(x));
+#else
+  int64_t i;
```
Comment (JanuszL) [inline include/dali/core/math_util.h]: Any reference to source of that magic?

File: include/dali/core/math_util.h (hunk 6 above)
Comment (JanuszL) [reply include/dali/core/math_util.h]: And justification why 2 iterations are sufficient?

File: include/dali/core/math_util.h [hunk 9]
```diff
@@ -53,6 +56,71 @@ DALI_HOST_DEV constexpr T clamp(const T &value, const T &lo, const T &hi) {
...
+  double x2, y;
+  x2 = x * 0.5;
+  y  = x;
+  i  = *(const int64_t*)&y;
+  i  = 0x5FE6EB50C7B537A9 - (i >> 1);
+  y  = *(const double *)&i;
+  y  = y * (1.5 - (x2 * y * y));
```
Comment (JanuszL) [inline include/dali/core/math_util.h]: Why 3 and not 2 or 4?

File: include/dali/core/math_util.h (hunk 6 above)
Comment (mzient) [reply include/dali/core/math_util.h]: https://github.com/id-Software/Quake-III-Arena/blob/master/code/game/q_math.c#L561

**Comment by mzient** [comment]
!build

File: include/dali/core/math_util.h (hunk 9 above)
Comment (mzient) [reply include/dali/core/math_util.h]: I measured it - it gave ~11 significant digits. 2 iterations weren't good enough.

File: include/dali/core/math_util.h (hunk 8 above)
Comment (mzient) [reply include/dali/core/math_util.h]: https://en.wikipedia.org/wiki/Fast_inverse_square_root

File: include/dali/core/math_util.h (hunk 7 above)
Comment (mzient) [reply include/dali/core/math_util.h]: https://en.wikipedia.org/wiki/Fast_inverse_square_root

File: include/dali/core/math_util.h (hunk 6 above)
Comment (mzient) [reply include/dali/core/math_util.h]: https://en.wikipedia.org/wiki/Fast_inverse_square_root
2 iterations give 6 significant digits.

File: include/dali/core/math_util.h (hunk 7 above)
Comment (JanuszL) [reply include/dali/core/math_util.h]: Adding that in comment would be helpful.

File: include/dali/core/math_util.h (hunk 8 above)
Comment (JanuszL) [reply include/dali/core/math_util.h]: Adding that in comment would be helpful.

File: include/dali/core/math_util.h (hunk 9 above)
Comment (JanuszL) [reply include/dali/core/math_util.h]: Adding that in comment would be helpful.

Comment (mzient) [reply include/dali/core/math_util.h]: Done.

File: include/dali/core/math_util.h (hunk 8 above)
Comment (mzient) [reply include/dali/core/math_util.h]: Done.

File: include/dali/core/math_util.h (hunk 7 above)
Comment (mzient) [reply include/dali/core/math_util.h]: Done.

File: include/dali/core/math_util.h (hunk 6 above)
Comment (mzient) [reply include/dali/core/math_util.h]: Comments added.

**Comment by mzient** [comment]
!build

File: include/dali/core/math_util.h [hunk 10]
```diff
@@ -56,64 +59,82 @@ DALI_HOST_DEV constexpr T clamp(const T &value, const T &lo, const T &hi) {
...
+  return rsqrt(static_cast<float>(x));
 #else
+  // No inverse square root intrinsic for double.
+  // Use bit-hack; faster than 1.0/sqrt(x).
+  // https://en.wikipedia.org/wiki/Fast_inverse_square_root
   int64_t i;
   double x2, y;
   x2 = x * 0.5;
   y  = x;
   i  = *(const int64_t*)&y;
```
Comment (klecki) [inline include/dali/core/math_util.h]: This is UB. :P

File: include/dali/core/geom/vec.h [hunk 11]
```diff
@@ -79,6 +79,46 @@ template <typename T>
 struct is_scalar : std::integral_constant<bool, !is_mat<T>::value && !is_vec<T>::value> {};
```
Comment (klecki) [inline include/dali/core/geom/vec.h]: Can you add template variable aliases for those?

Comment (mzient) [reply include/dali/core/geom/vec.h]: Done.

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: include/dali/core/geom/vec.h [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+    decltype(v[0]*v[0] + v[0]*v[0]) ret = v[0]*v[0];
+    for (size_t i = 1; i < N; i++)
+      ret += v[i]*v[i];
+    return ret;
+  }
+
+  /// @brief Calculates Euclidean length of the vector.
+  DALI_HOST_DEV inline auto length() const {
+#ifdef __CUDA_ARCH__
+    return sqrtf(length_square());
```
Comment (JanuszL) [inline include/dali/core/geom/vec.h]: Inconsistent indentation

File: include/dali/core/geom/vec.h [hunk 13]
```diff
@@ -0,0 +1,572 @@
...
+}
+
+/// @brief Calculates `z` coordinate of a cross product of two 2D vectors
+template <typename T, typename U>
+DALI_HOST_DEV
+constexpr auto cross(const vec<2, T> &a, const vec<2, U> &b) {
+  return a.x * b.y - b.x * a.y;
+}
+
+#define DEFINE_ELEMENTIWSE_VEC_BIN_OP(op)\
```
Comment (JanuszL) [inline include/dali/core/geom/vec.h]: Please align `\`

Comment (mzient) [reply include/dali/core/geom/vec.h]: Done.

File: include/dali/core/geom/vec.h (hunk 12 above)
Comment (mzient) [reply include/dali/core/geom/vec.h]: Fixed.

**Comment by mzient** [comment]
!build

**Pull request merged**